File: fm-domain/src/main/java/io/zeitwert/dddrive/ddd/adapter/api/jsonapi/dto/AggregateDtoBase.java
package io.zeitwert.dddrive.ddd.adapter.api.jsonapi.dto;

import com.fasterxml.jackson.annotation.JsonIgnore;
import dddrive.ddd.core.model.Aggregate;
import io.crnk.core.resource.annotations.JsonApiField;
import io.crnk.core.resource.annotations.JsonApiId;
import io.zeitwert.dddrive.ddd.adapter.api.jsonapi.AggregateDto;
import io.zeitwert.dddrive.ddd.adapter.api.jsonapi.AggregateDtoAdapter;
import io.zeitwert.dddrive.ddd.api.rest.dto.EnumeratedDto;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

@Data
@SuperBuilder
@NoArgsConstructor
public abstract class AggregateDtoBase<A extends Aggregate> implements AggregateDto<A> {

	public static final String CalculationOnlyOperation = "calculationOnly";

	@JsonIgnore
	private AggregateDtoAdapter<?, ?> adapter;
	@JsonApiId
	private Integer id;
	private String caption;
	// Read: for orderbooks, write: for creation
	private EnumeratedDto tenant;
	// Read: for orderbooks, write: for updates
	private EnumeratedDto owner;

	public AggregateDtoAdapter<?, ?> getAdapter() {
		return this.adapter;
	}

	public abstract AggregateMetaDto getMeta();

	// For explicit filtering in SaaS session
	@JsonApiField(readable = false, filterable = true)
	public Integer getTenantId() {
		return null;
	}

	@JsonApiField(readable = false, filterable = true)
	public String getSearchText() {
		return null;
	}

}

File: fm-domain/src/main/java/io/zeitwert/dddrive/ddd/adapter/api/jsonapi/base/AggregateDtoBase.kt
package io.zeitwert.dddrive.ddd.adapter.api.jsonapi.base

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonIgnore
import dddrive.app.ddd.model.Aggregate
import io.crnk.core.resource.annotations.JsonApiId
import io.crnk.core.resource.annotations.JsonApiMetaInformation
import io.zeitwert.dddrive.ddd.adapter.api.jsonapi.AggregateDto

/**
 * Base class for generic JSON API resources.
 *
 * Uses @JsonAnyGetter/@JsonAnySetter for dynamic attribute handling,
 * allowing properties to be serialized without explicit field declarations.
 *
 * Subclasses only need to add the @JsonApiResource annotation with the
 * appropriate type and resourcePath.
 */
abstract class AggregateDtoBase<A : Aggregate> : AggregateDto<A> {

	@JsonApiMetaInformation
	override var meta: MutableMap<String, Any?>? = null

	@JsonIgnore
	private val attributes: MutableMap<String, Any?> = mutableMapOf()

	@JsonIgnore
	private val relations: MutableMap<String, Any?> = mutableMapOf()

	@JsonApiId
	override var id: String? = null
		get() = attributes["id"] as? String
		set(value) {
			attributes["id"] = value
			field = value
		}

	override fun containsKey(key: String): Boolean = attributes.containsKey(key)

	override operator fun get(key: String): Any? = attributes[key]

	@JsonAnyGetter
	fun any(): Map<String, Any?> = attributes

	@JsonAnySetter
	override operator fun set(
		key: String,
		value: Any?,
	) {
		attributes[key] = value
	}

	override fun hasRelation(name: String): Boolean = relations.containsKey(name)

	override fun setRelation(
		name: String,
		value: Any?,
	) {
		relations[name] = value
	}

	override fun getRelation(name: String): Any? = relations[name]

	@Suppress("UNCHECKED_CAST")
	override fun hasOperation(name: String): Boolean {
		val operations = meta?.get("operations") as List<String>? ?: emptyList()
		return operations.contains(name)
	}

	override fun toString() = "${javaClass.simpleName}[$id] meta: $meta $attributes"

}
